src/Symfony/Component/Cache/Adapter/TagAwareAdapter.php

```diff
@@ -363,39 +364,39 @@ private function getTagVersions(array $tagsByKey, bool $persistTags): array
         $tags = [];
         foreach ($tagVersions as $tag => $version) {
             $tags[$tag.static::TAGS_PREFIX] = $tag;
+            if ($fetchTagVersions) {
+                continue;
+            }
             $knownTagVersion = $this->knownTagVersions[$tag] ?? [0, null];
-            if ($fetchTagVersions || $knownTagVersion[1] !== $version || $now - $knownTagVersion[0] >= $this->knownTagVersionsTtl) {
+            if ($knownTagVersion[1] !== $version || $now >= $knownTagVersion[0]) {
                 // reuse previously fetched tag versions up to the ttl
                 $fetchTagVersions = true;
             }
-            unset($this->knownTagVersions[$tag]); // For LRU tracking
-            if ([0, null] !== $knownTagVersion) {
-                $this->knownTagVersions[$tag] = $knownTagVersion;
-            }
         }
 
         if (!$fetchTagVersions) {
             return $tagVersions;
         }
 
+        if (self::MAX_NUMBER_OF_KNOWN_TAG_VERSIONS < \count($this->knownTagVersions)) {
+            $this->knownTagVersions = array_filter($this->knownTagVersions, static fn ($v) => $now < $v[0]);
+        }
+
         $newTags = [];
         $newVersion = null;
+        $expiration = $now + $this->knownTagVersionsTtl;
         foreach ($this->tags->getItems(array_keys($tags)) as $tag => $version) {
-            if (!$version->isHit()) {
+            if ($persistTags && !$version->isHit()) {
```

Comment:
to my understanding, this will force extra round-trips to the backend.
Here are the steps, assuming the cache is empty:
1. get items
2. get their tag versions for validation (since cache is empty)
3. get their tag versions again, for populating items (cache is always ignored)
4. send items

While the previous process was:
1. get items
2. get their tag versions for validation (since cache is empty)
3. get their tag versions again, for populating items, but only for tags not already fetched at the previous step!
4. send items

For new/expired items, both processes are equivalent, but for items that invalidated (aka tag versions don't match), doing a round-trip on step 3. was a lot less likely (occurs only when the updated items has new tags.)

That's why I'm claiming that the proposed process is going to reduce perf by doing more round-trips. I don't get why you claim that your proposal here is going to be more performant. Can you elaborate?

Reply:
Thank you for detailed comment!

As I wrote earlier, the change in this line doesn't change the number of round-trips. This additional condition prevents false-positive _reads_ when tag was invalidated. I suppose that if tag doesn't exist, we should return null instead of just generated version since it may repeat (one of) previous version(s). This is 2. in the list of changes.

Forcing the adapter to fetch tags on commits doesn't lead to extra work (at the moment), it just makes the intention explicit. Please notice that for items being persisted tags are extracted from their metadata, hence $tagsByKey contains tag names `['item1' => ['tag123' => 'tag123', ...], ...]`. The only chance to skip fetching was when tag name happens to be equal to its random version. `$forceFetching` just eliminates this possibility. And instructs to skip unnecessary comparisons. This is 1. in the list of changes.

In fact, this adapter has been never using known tag versions for item commits. 
https://github.com/symfony/symfony/commit/212c469ebfbbd2d38981eef123b75cb7e4159cc5#diff-47997875179d5ba246b68c6ca2aac323cd05de24661476198f2758d2b3094ef7R333
And in this PR I tried to keep this behavior unchanged.

But I like your view on how it could be improved.

I've updated `getTagVersions()` to exclude a round-trip on step 3. The algorithm became a bit more complicated than the current implementation but I hope it's worth it.

Reply:
I am afraid, postponing persisting new version of tags is not the best strategy. From the perspective of overall performance, it's better to persist tags as soon as possible (like it was before the upgrade) and reread them before committing to let all concurrent clients use the same (the latest) version of tags. Whith late persisting every client will use its own versions, and all of them but the last one will have to recompute their items.

Reply:
After some experiments I found current fetching strategy (to fetch all requested tags before every item commit, and to persist absent tags only before item commits) optimal.

The only use case when persisting absent tags at every tag request demonstrates better performance (less item computations) is when $knownTagVersionsTtl = 0 and PSR-6 methods are used.

When we do not fetch tags before item commit, we risk either attaching stale tag versions (if we take them from cache) or overwriting already persisted versions which were attached to other items (if we perssst them after a miss on `getItem()`). Both paths leads to appearence of invalid items in the pool with subsequent recomputations. `get()` and `LockRegistry` are powerless in these cases.

Fetching only uknown/expired tags also potentially increases the number of tag reads and writes since they are processed in smaller sets, or even one by one.

Reply:
Works for me, thanks for the details.